src/resolve/resolved-dns-cache.c

```diff
@@ -623,9 +623,13 @@ static int dns_cache_put_negative(
                 calculate_until_valid(soa, dns_answer_min_ttl(answer), nsec_ttl, timestamp, true);
 
         if (i->type == DNS_CACHE_NXDOMAIN) {
+                const char *name = dns_resource_key_name(key);
+                if (dns_name_endswith(name, "resolver.arpa"))
+                        return 0;
```

Comment:
can you split this out into a helper and add a comment explaining the story (i.e. the stuff from the commit msg should be a comment).

call the helper `bool cache_nxdomain_for(DnsResourceKey *key)` or so?

(somewhere in the back of my head i have the suspicion this helper will need another check sooner or later, that refuses to cache NXDOMAIN on AAAA to deal with this mess: #22575 #20548)

Reply:
Wow that's a bother.

Anyway I did rework this but not exactly as requested. PTAL.

Reply:
Regarding broken AAAA responses...

Idk maybe just tell people to use Cache=no-negative and deal with it. Perhaps it could be mitigated slightly by trying to avoid "contradictory" caching, so if we send A/AAAA at the same time, and the answer contradict (one NXDOMAIN, one NOERROR), we don't cache the NXDOMAIN.

Reply:
i guess we could implement a logic that when the cache is updated and we already have a valid cached entry that is contradicting it we mark the cache entry in a new CONTRADICTING state, and then report that back to the transaction logic whenever the cache is queried next. It would then basically be treated as a hint to always ask the server again.